### PR TITLE
UPDATED: sqlc.narg() - allow user-specified nullability

### PR DESCRIPTION
**Pull request by skabbes**
EDITED from original PR after discussion here:  https://github.com/kyleconroy/sqlc/issues/1525

**Exec Summary**
I implemented user-specified nullability with `sqlc.narg`

 ```sql
sqlc.narg('param') -- regardless of whether this parameter was inferred to be nullable or not, it will be nullable in the generated code.
 ```

**Details**
Most of the work here came from building types to represent set of named parameters, which tracks the name and nullability status of any parameter (inferred or user defined).

`named.Param`
This type can track its nullability from the sqlc engine (inferred nullability) separately from any user-specified nullability, and always prefers the user-specified version to the inferred one.

`named.ParamSet`
This type is essentially a collection of params, and helps build up parameters, and encapsulate their positional representations as well.  As you can see - this type significantly cleaned up `compiler/resolve.go`

**Next Steps**
This code works, is tested. I think a code review would be great, and maybe a pointer to how to update the docs (or maybe the doc repos are still private).

### CODE REVIEW

File: internal/endtoend/testdata/sqlc_arg/mysql/query.sql [hunk 1]
```diff
@@ -1,7 +1,13 @@
...
 SELECT name FROM foo WHERE name = sqlc.arg(slug);
 
 /* name: FuncParamString :many */
 SELECT name FROM foo WHERE name = sqlc.arg('slug');
+
+/* name: FuncParamStringOptional :exec */
+UPDATE foo SET name = coalesce(sqlc.arg('slug?'), name);
+
+/* name: FuncParamStringRequired :exec */
+UPDATE foo SET description = sqlc.arg('slug!');
```
Comment (skabbes) [inline internal/endtoend/testdata/sqlc_arg/mysql/query.sql:13]: Example usage of sqlc.arg with user-defineable nullability.

File: internal/sql/rewrite/parameters.go [hunk 2]
```diff
@@ -41,55 +41,61 @@ func isNamedParamSignCast(node ast.Node) bool {
...
+	last := name[len(name)-1]
+	if last == '!' {
+		return named.NewUserDefinedParam(name[:len(name)-1], true)
+	}
+
+	if last == '?' {
+		return named.NewUserDefinedParam(name[:len(name)-1], false)
+	}
+
+	return named.NewUnspecifiedParam(name)
```
Comment (skabbes) [inline internal/sql/rewrite/parameters.go:60]: This is the core logic for the nullability options, I'm happy to rip this out and replace it with anything else.

File: internal/sql/rewrite/parameters.go [hunk 3]
```diff
@@ -41,55 +41,61 @@ func isNamedParamSignCast(node ast.Node) bool {
...
-				if _, found := args[param]; !found {
-					args[param] = []int{argn}
-				} else {
-					args[param] = append(args[param], argn)
-				}
-				cr.Replace(&ast.ParamRef{
-					Number:   argn,
-					Location: fun.Location,
-				})
-			}
```
Comment (skabbes) [inline internal/sql/rewrite/parameters.go:81]: _aside_: encapsulating the positional tracking into `named.ParamSet` significantly cleaned up this code.

File: internal/source/code.go [hunk 4]
```diff
@@ -54,25 +54,31 @@ func Mutate(raw string, a []Edit) (string, error) {
 	if len(a) == 0 {
```
Comment (skabbes) [inline internal/source/code.go:54]: I was having a bit of trouble with this, so I tested the heck out of it and these were the resulting edits.

**Comment by skabbes** [comment]
@kyleconroy I went ahead and implemented `sqlc.narg` like we discussed, could you take a look at you earliest convenience?

File: internal/source/code.go (hunk 4 above)
Comment (kyleconroy) [reply internal/source/code.go:54]: The edit code is super hacky so I'm not surprised you needed to add all these tests. Hopeful in the future we can get rid of all of this code and generate SQL using an AST.

**Review by kyleconroy — APPROVED** [review]

**Comment by kyleconroy** [comment]
It's a lot to ask, but can you also add docs to https://docs.sqlc.dev/en/stable/howto/named_parameters.html in a separate PR? Want to make sure that people know this option exists.

**Comment by skabbes** [comment]
ah, I didn't look hard enough for the docs. But yes I can do that.

**Comment by skabbes** [comment]
Added here: https://github.com/kyleconroy/sqlc/pull/1579

**Comment by skabbes** [comment]
Just kidding, one that actually passes CI here: https://github.com/kyleconroy/sqlc/pull/1580

**Pull request merged**